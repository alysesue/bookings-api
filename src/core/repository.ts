import { Connection, Repository } from "typeorm";
import { Inject } from "typescript-ioc";
import { DbConnection } from "./db.connection";
import { logger } from "mol-lib-common/debugging/logging/LoggerV2";

export abstract class RepositoryBase<T> {

	@Inject
	private connection: DbConnection;

	private readonly modelType;

	protected constructor(modelType) {
		this.modelType = modelType;
	}

	protected async getRepository(): Promise<Repository<T>> {
		try {
			const conn = await this.connection.getConnection();
			return conn.getRepository<T>(this.modelType);
		} catch (e) {
			logger.error("ServiceProviderRepository::connection::error", e);
			throw e;
		}
	}

	protected async getConnection(): Promise<Connection> {
		return this.connection.getConnection();
	}
}

export enum QueryAccessType {
	Read = 1,
	Write
}
