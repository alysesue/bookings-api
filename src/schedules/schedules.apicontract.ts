import { Weekday } from "../enums/weekday";
import { Schedule } from "../models";

export class ScheduleRequest {
	/**
	 * @maxLength 100
	 */
	public name: string;
	public slotsDurationInMin: number;
	public weekdaySchedules: WeekDayScheduleContract[];
}

export class WeekDayScheduleContract {
	public weekday: Weekday;
	public hasSchedule: boolean;
	public openTime?: string;
	public closeTime?: string;
}

export class ScheduleResponse {
	public id: number;
	public name: string;
	public slotsDurationInMin: number;
	public weekdaySchedules: WeekDayScheduleContract[];
}
