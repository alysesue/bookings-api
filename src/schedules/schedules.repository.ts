import { Inject, Singleton } from 'typescript-ioc';

import { DbConnection } from '../core/db.connection';
import { Schedule } from '../models/Schedule';
import { DeleteResult, Repository } from "typeorm";

@Singleton
export class SchedulesRepository {
	@Inject
	private connection: DbConnection;

	public async getScheduleById(id: number): Promise<Schedule> {
		return (await this.getRepository()).findOne({ id });
	}

	public async getScheduleByName(name: string): Promise<Schedule> {
		return (await this.getRepository()).findOne({ name });
	}

	public async setSchedule(timeslot: Schedule): Promise<Schedule> {
		return (await this.getRepository()).save(timeslot);
	}

	public async deleteSchedule(timeslotId: number): Promise<DeleteResult> {
		return (await this.getRepository()).delete(timeslotId);
	}

	private async getRepository(): Promise<Repository<Schedule>> {
		const conn = await this.connection.getConnection();
		return conn.getRepository(Schedule);
	}
}
