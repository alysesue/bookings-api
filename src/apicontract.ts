import { IPagedEntities } from './core/pagedEntities';

export class ApiData<T> {
	public data: T;
	constructor(typedData: T) {
		this.data = typedData;
	}
}

export class FailedRecord<Request, Reason> {
	public request: Request;
	public reason: Reason;
	constructor(request: Request, reason: Reason) {
		this.request = request;
		this.reason = reason;
	}
}
export class ApiDataBulk<T, FailedRecords> {
	public created: T;
	public failed: FailedRecords;
	constructor(typedData: T, failedRecords: FailedRecords) {
		this.created = typedData;
		this.failed = failedRecords;
	}
}

export type PagingRequest = {
	page: number;
	limit: number;
	maxId?: number;
};

export class ApiPagingInfo {
	public page: number;
	public limit: number;
	public total: number;
	public maxId: number;
	public outdatedMaxId: boolean;
	public hasMore: boolean;
}

export class ApiPagedData<T> {
	public data: T[];
	public paging: ApiPagingInfo;
}

export type MapFunction<TInput, TOutput> = (input: TInput) => TOutput;
export class ApiDataFactory {
	private constructor() {}
	public static create<T>(data: T): ApiData<T> {
		return new ApiData(data);
	}

	public static createBulk<T, F>(created: T, failedBookings: F): ApiDataBulk<T, F> {
		return new ApiDataBulk(created, failedBookings);
	}

	private static createPagingInfo(pagedEntities: IPagedEntities<any>): ApiPagingInfo {
		const info = new ApiPagingInfo();
		info.page = pagedEntities.page;
		info.limit = pagedEntities.limit;
		info.total = pagedEntities.total;
		info.maxId = pagedEntities.maxId;
		info.outdatedMaxId = pagedEntities.outdatedMaxId;
		info.hasMore = pagedEntities.hasMore;
		return info;
	}

	public static createPaged<TInput, TOutput>(
		pagedEntities: IPagedEntities<TInput>,
		mapFunction: MapFunction<TInput, TOutput>,
	): ApiPagedData<TOutput> {
		const pagedData = new ApiPagedData<TOutput>();
		pagedData.data = pagedEntities.entries.map(mapFunction);
		pagedData.paging = ApiDataFactory.createPagingInfo(pagedEntities);
		return pagedData;
	}
}
