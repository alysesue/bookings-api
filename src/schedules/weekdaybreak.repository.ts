import { Scope, Scoped } from 'typescript-ioc';
import { WeekDayBreak } from '../models';
import { DeleteResult, In } from "typeorm";
import { RepositoryBase } from '../core/repository';

@Scoped(Scope.Request)
export class WeekDayBreakRepository extends RepositoryBase<WeekDayBreak> {

	constructor() {
		super(WeekDayBreak);
	}

	public async getBreaksForSchedules(scheduleIds: number[]): Promise<WeekDayBreak[]> {
		return (await this.getRepository()).find({
			where: {
				scheduleId: In(scheduleIds)
			}
		});
	}

	public async deleteBreaksForSchedule(scheduleId: number): Promise<DeleteResult> {
		const connection = (await this.getConnection());
		const query = connection
			.createQueryBuilder()
			.delete()
			.from(WeekDayBreak)
			.where("scheduleId = :scheduleId", { scheduleId });

		return query.execute();
	}

	public async save(entities: WeekDayBreak[]): Promise<WeekDayBreak[]> {
		return (await this.getRepository()).save(entities);
	}
}
