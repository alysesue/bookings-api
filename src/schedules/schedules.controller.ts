import { Body, Controller, Delete, Get, Path, Post, Put, Route, SuccessResponse, Tags } from 'tsoa';
import { ScheduleRequest, ScheduleResponse } from "./schedules.apicontract";
import { Inject } from "typescript-ioc";
import { SchedulesService } from "./schedules.service";

@Route('api/v1/schedules')
@Tags('Schedules')
export class SchedulesController extends Controller {
	@Inject
	private scheduleService: SchedulesService;

	@Post('')
	@SuccessResponse(201, 'Created')
	public async createSchedule(@Body() timeslot: ScheduleRequest): Promise<ScheduleResponse> {
		return await this.scheduleService.createSchedule(timeslot);
	}

	@Get("")
	public async getSchedules(): Promise<ScheduleResponse[]> {
		return await this.scheduleService.getSchedules();
	}

	@Put('{id}')
	@SuccessResponse(200, 'Updated')
	public async updateSchedule(@Path() id: number, @Body() timeslot: ScheduleRequest): Promise<ScheduleResponse> {
		return await this.scheduleService.updateSchedule(id, timeslot);
	}

	@Delete('{id}')
	@SuccessResponse(200, 'Deleted')
	public async deleteSchedule(@Path() id: number): Promise<any> {
		return await this.scheduleService.deleteSchedule(id);
	}
}
