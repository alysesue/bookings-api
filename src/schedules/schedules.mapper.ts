import { WeekdayList } from "../enums/weekday";
import { Schedule, TimeOfDay, WeekDaySchedule } from "../models";
import { ScheduleRequest, ScheduleResponse, WeekDayScheduleContract } from './schedules.apicontract';
import { groupByKeyLastValue } from '../tools/collections';
import { getErrorResult, isErrorResult, OptionalResult } from '../errors';

export const mapToEntity = (contract: ScheduleRequest, entity: Schedule): OptionalResult<Schedule, string[]> => {
	entity.name = contract.name;
	entity.slotsDurationInMin = contract.slotsDurationInMin;
	const errors: string[] = [];

	const weekDaysContract = groupByKeyLastValue(contract.weekdaySchedules || [], w => w.weekday);

	if (!entity.weekdaySchedules || entity.weekdaySchedules.length === 0) {
		entity.weekdaySchedules = WeekdayList.map(day => new WeekDaySchedule(day));
	}

	for (const daySchedule of entity.weekdaySchedules) {
		if (weekDaysContract.has(daySchedule.weekDay)) {
			const dayContract = weekDaysContract.get(daySchedule.weekDay);

			const dayResult = setDayContractEntity(dayContract, daySchedule);
			if (isErrorResult(dayResult)) {
				errors.push(...getErrorResult(dayResult));
			}
		}
	}

	if (errors.length > 0) {
		return {
			errorResult: errors
		};
	}
	return {
		result: entity
	};
};

const setDayContractEntity = (contract: WeekDayScheduleContract, entity: WeekDaySchedule): OptionalResult<WeekDaySchedule, string[]> => {
	entity.hasSchedule = contract.hasSchedule;
	const errors: string[] = [];

	try {
		entity.openTime = TimeOfDay.parse(contract.openTime);
	} catch (e) {
		errors.push(e.message);
	}

	try {
		entity.closeTime = TimeOfDay.parse(contract.closeTime);
	} catch (e) {
		errors.push(e.message);
	}

	if (errors.length > 0) {
		return {
			errorResult: errors
		};
	}

	return {
		result: entity
	};
};

export const mapToResponse = (template: Schedule): ScheduleResponse => {
	const response = new ScheduleResponse();
	response.id = template.id;
	response.name = template.name;
	response.slotsDurationInMin = template.slotsDurationInMin;
	response.weekdaySchedules = template.weekdaySchedules?.map(w => mapDayScheduleToResponse(w)) || [];

	return response;
};

export const mapDayScheduleToResponse = (daySchedule: WeekDaySchedule): WeekDayScheduleContract => {
	const dayContract = new WeekDayScheduleContract();
	dayContract.weekday = daySchedule.weekDay;
	dayContract.hasSchedule = daySchedule.hasSchedule;
	dayContract.openTime = daySchedule.openTime?.toJSON();
	dayContract.closeTime = daySchedule.closeTime?.toJSON();

	return dayContract;
};
