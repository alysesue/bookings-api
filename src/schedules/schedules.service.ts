import { ErrorCodeV2, MOLErrorV2 } from "mol-lib-api-contract";
import { Inject, Singleton } from 'typescript-ioc';
import { DeleteResult } from "typeorm";
import { SchedulesRepository } from "./schedules.repository";
import { Schedule } from '../models/Schedule';
import { ScheduleRequest, ScheduleResponse } from "./schedules.apicontract";
import { mapToEntity, mapToResponse } from './schedules.mapper';
import { getErrorResult, isErrorResult, OptionalResult } from '../errors';

@Singleton
export class SchedulesService {
	@Inject
	private schedulesRepository: SchedulesRepository;

	private validate(schedule: Schedule) {
		const validations = Array.from(schedule.validateSchedule());
		if (validations.length > 0) {
			const response = validations.map(val => val.message);
			throw new MOLErrorV2(ErrorCodeV2.SYS_INVALID_PARAM).setResponseData(response);
		}
	}

	private mapToEntityAndValidate(template: ScheduleRequest, schedule: Schedule) {
		const mapped = mapToEntity(template, schedule);
		if (isErrorResult(mapped)) {
			const errorResult = getErrorResult(mapped);
			throw new MOLErrorV2(ErrorCodeV2.SYS_INVALID_PARAM).setResponseData(errorResult);
		}

		const validations = Array.from(schedule.validateSchedule());
		if (validations.length > 0) {
			const response = validations.map(val => val.message);
			throw new MOLErrorV2(ErrorCodeV2.SYS_INVALID_PARAM).setResponseData(response);
		}

		return schedule;
	}

	public async createSchedule(template: ScheduleRequest): Promise<ScheduleResponse> {
		const newSchedule = this.mapToEntityAndValidate(template, new Schedule());
		const templateSet = (await this.schedulesRepository.saveSchedule(newSchedule));
		return mapToResponse(templateSet);
	}

	public async updateSchedule(id: number, template: ScheduleRequest): Promise<ScheduleResponse> {
		const existingSchedule = await this.schedulesRepository.getScheduleById(id);
		if (!existingSchedule) {
			throw new MOLErrorV2(ErrorCodeV2.SYS_NOT_FOUND).setMessage('Schedule not found.');
		}
		let schedule = this.mapToEntityAndValidate(template, existingSchedule);

		schedule = (await this.schedulesRepository.saveSchedule(schedule));
		return mapToResponse(schedule);
	}

	public async getSchedules(): Promise<ScheduleResponse[]> {
		const schedules = await this.schedulesRepository.getSchedules();
		return schedules.map(s => mapToResponse(s));
	}

	public async getSchedule(id: number): Promise<ScheduleResponse> {
		const schedule = await this.schedulesRepository.getScheduleById(id);
		return mapToResponse(schedule);
	}

	public async deleteSchedule(id: number): Promise<DeleteResult> {
		return await this.schedulesRepository.deleteSchedule(id);
	}
}
