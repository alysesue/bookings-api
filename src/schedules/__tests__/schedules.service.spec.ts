import { SchedulesService } from "../schedules.service";
import { ScheduleRequest, WeekDayScheduleContract } from "../schedules.apicontract";
import { SchedulesRepository } from "../schedules.repository";
import { Container } from "typescript-ioc";
import { Schedule } from "../../models/Schedule";
import { mapToEntity } from '../schedules.mapper';
import { Weekday } from "../../enums/weekday";

const scheduleRequestCommon = {
	name: 'schedule',
	slotsDurationInMin: 60,
	weekdaySchedules: [
		{ weekday: Weekday.Monday, hasSchedule: true, openTime: '11:23', closeTime: '12:23' } as WeekDayScheduleContract
	]
} as ScheduleRequest;

const scheduleCommon: Schedule = mapToEntity(scheduleRequestCommon, new Schedule());

const getSchedules = jest.fn().mockImplementation(() => Promise.resolve([scheduleCommon]));
const getScheduleById = jest.fn().mockImplementation(() => Promise.resolve(scheduleCommon));
const getScheduleByName = jest.fn().mockImplementation(() => Promise.resolve(scheduleCommon));
const saveSchedule = jest.fn().mockImplementation(() => Promise.resolve(scheduleCommon));
const deleteSchedule = jest.fn().mockImplementation(() => Promise.resolve(undefined));
const MockSchedulesRepository = jest.fn().mockImplementation(() => ({
	getSchedules,
	saveSchedule,
	getScheduleById,
	getScheduleByName,
	deleteSchedule
}));

describe('Schedules  template services ', () => {
	let schedulesService: SchedulesService;
	beforeAll(() => {
		Container.bind(SchedulesRepository).to(MockSchedulesRepository);
		schedulesService = Container.get(SchedulesService);
	});
	beforeEach(() => {
		jest.clearAllMocks();
	});

	it('should throw error because firstSlotStartTimeInHHmm have wrong format', async () => {
		const schedulesRequest: ScheduleRequest = {
			name: 'schedule',
			slotsDurationInMin: 5,
			weekdaySchedules: [
				{ weekday: Weekday.Monday, hasSchedule: true, openTime: '2323', closeTime: '12:23' } as WeekDayScheduleContract
			]
		} as ScheduleRequest;

		try {
			await schedulesService.createSchedule(schedulesRequest);
		} catch (e) {
			expect(e.message).toBe("Not valid format for firstSlotStartTimeInHHmm: 2323");
		}
		expect(saveSchedule).toBeCalledTimes(0);
	});

	it('should throw error because lastSlotEndTimeInHHmm have wrong format', async () => {
		const schedulesRequest: ScheduleRequest = {
			name: 'schedule',
			slotsDurationInMin: 5,
			weekdaySchedules: [
				{ weekday: Weekday.Monday, hasSchedule: true, openTime: '23:23', closeTime: '11:73' } as WeekDayScheduleContract
			]
		} as ScheduleRequest;

		try {
			await schedulesService.createSchedule(schedulesRequest);
		} catch (e) {
			expect(e.message).toBe("Not valid format for lastSlotEndTimeInHHmm: 11:73");
		}
		expect(saveSchedule).toBeCalledTimes(0);
	});

	it('should throw error because firstSlotStartTimeInHHmm > lastSlotEndTimeInHHmm', async () => {
		const schedulesRequest: ScheduleRequest = {
			name: 'schedule',
			slotsDurationInMin: 5,
			weekdaySchedules: [
				{ weekday: Weekday.Monday, hasSchedule: true, openTime: '23:23', closeTime: '11:23' } as WeekDayScheduleContract
			]
		} as ScheduleRequest;

		try {
			await schedulesService.createSchedule(schedulesRequest);
		} catch (e) {
			expect(e.message).toBe("firstSlotStartTimeInHHmm=23:23 > lastSlotEndTimeInHHmm=11:23");
		}
		expect(saveSchedule).toBeCalledTimes(0);

	});

	it('should throw error because slotsDurationInMin < lastSlotEndTimeInHHmm - firstSlotStartTimeInHHmm ', async () => {
		const schedulesRequest: ScheduleRequest = {
			name: 'schedule',
			slotsDurationInMin: 65,
			weekdaySchedules: [
				{ weekday: Weekday.Monday, hasSchedule: true, openTime: '11:23', closeTime: '12:23' } as WeekDayScheduleContract
			]
		} as ScheduleRequest;

		try {
			await schedulesService.createSchedule(schedulesRequest);
		} catch (e) {
			expect(e.message).toBe("slotsDurationInMin=65 < (lastSlotEndTimeInHHmm-firstSlotStartTimeInHHmm)=60");
		}
		expect(saveSchedule).toBeCalledTimes(0);

	});

	it('should create new Schedule ', async () => {
		await schedulesService.createSchedule(scheduleRequestCommon);
		expect(saveSchedule).toBeCalledTimes(1);
	});

	it('should update the template', async () => {
		const template = await schedulesService.updateSchedule(1, scheduleRequestCommon);

		expect(saveSchedule).toBeCalled();
		expect(getScheduleByName).toBeCalled();
		expect(template.name).toStrictEqual(scheduleRequestCommon.name);
	});

	it('should call delete repository', async () => {
		await schedulesService.deleteSchedule(3);
		expect(deleteSchedule).toBeCalled();
	});

});
