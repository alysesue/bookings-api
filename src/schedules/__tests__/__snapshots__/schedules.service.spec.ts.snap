// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Schedules  template services  should throw error because firstSlotStartTimeInHHmm > lastSlotEndTimeInHHmm 1`] = `
Array [
  "Close time [[object Object]] must be greater than open time [[object Object]]",
]
`;

exports[`Schedules  template services  should throw error because firstSlotStartTimeInHHmm have wrong format 1`] = `
Array [
  "Value 2323 is not a valid time.",
  "Value 25:25 is not a valid time.",
]
`;

exports[`Schedules  template services  should throw error because lastSlotEndTimeInHHmm have wrong format 1`] = `
Array [
  "Value 11:73 is not a valid time.",
]
`;

exports[`Schedules  template services  should throw error because slotsDurationInMin < lastSlotEndTimeInHHmm - firstSlotStartTimeInHHmm  1`] = `
Array [
  "The interval between open and close times [[object Object] â€” [object Object]] must be greater that slot duration [65]",
]
`;
